File: exercism/java/triangle/src/main/java/Triangle.java
class Triangle {
  double _side1;
  double _side2;
  double _side3;

  Triangle(double side1, double side2, double side3) throws TriangleException {
    if ((side1 <= 0) || (side2 <= 0) || (side3 <= 0)) {
      throw new TriangleException("lengths must be positive");
    } else if ((side1 + side2 < side3) || (side1 + side3 < side2) || (side2 + side3 < side1)) {
      throw new TriangleException("fails triangle inequality");
    } else {
      _side1 = side1;
      _side2 = side2;
      _side3 = side3;
    }
  }

  boolean isEquilateral() {
    return (_side1 == _side2) && (_side2 == _side3);
  }

  boolean isIsosceles() {
    return (_side1 == _side2) || (_side1 == _side3) || (_side2 == _side3);
  }

  boolean isScalene() {
    return !(isEquilateral() || isIsosceles());
  }
}

File: exercism/ocaml/triangle/triangle.ml
open Base

let ok a b c = a > 0 && b > 0 && c > 0 && a + b >= c && a + c >= b && b + c >= a
let is_equilateral a b c = ok a b c && equal a b && equal b c
let is_isosceles a b c = ok a b c && (equal a b || equal a c || equal b c)
let is_scalene a b c = ok a b c && a <> b && a <> c && b <> c
